CodePushDownloadHandler.m

```diff
@@ -0,0 +1,79 @@
+#import "CodePush.h"
+
+@implementation CodePushDownloadHandler
+
+- (id)init:(NSString *)downloadFilePath
+progressCallback:(void (^)(long, long))progressCallback
+doneCallback:(void (^)())doneCallback
+failCallback:(void (^)(NSError *err))failCallback {
+    self.outputFileStream = [NSOutputStream outputStreamToFileAtPath:downloadFilePath
+                                                              append:NO];
+    self.receivedContentLength = 0;
+    self.progressCallback = progressCallback;
+    self.doneCallback = doneCallback;
+    self.failCallback = failCallback;
+    return self;
+}
+
+-(void)download:(NSString*)url {
+    NSURLRequest *request = [NSURLRequest requestWithURL:[NSURL URLWithString:url]
+                                             cachePolicy:NSURLRequestUseProtocolCachePolicy
+                                         timeoutInterval:60.0];
+    
+    NSURLConnection *connection = [[NSURLConnection alloc] initWithRequest:request
+                                                                  delegate:self
+                                                          startImmediately:NO];
+    [connection scheduleInRunLoop:[NSRunLoop mainRunLoop]
+                          forMode:NSDefaultRunLoopMode];
+    [connection start];
+}
+
+#pragma mark NSURLConnection Delegate Methods
+
+- (NSCachedURLResponse *)connection:(NSURLConnection *)connection
+                  willCacheResponse:(NSCachedURLResponse*)cachedResponse {
+    // Return nil to indicate not necessary to store a cached response for this connection
+    return nil;
+}
+
+-(void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
+    self.expectedContentLength = response.expectedContentLength;
+    [self.outputFileStream open];
+}
+
+-(void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
+    self.receivedContentLength = self.receivedContentLength + [data length];
+    
+    NSInteger bytesLeft = [data length];
+    
+    do {
+        NSInteger bytesWritten = [self.outputFileStream write:[data bytes]
```

Comment:
I have added some tests cases in the sample app, look for Examples/CodePushDemoApp/CodePushDemoAppTests/DownloadProgressTests.

They use the react packager (for ease of running, since it starts up automatically when u run the test project) to act as a server to serve 3 pacakges, one 100kb, one 2mb, one 50mb.